{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "49ea04a1",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1683296",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/henrysun_1/Desktop/Duke/PhD 2025-2026/callinectes_co1\n"
     ]
    }
   ],
   "source": [
    "import Bio \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "print(os.getcwd())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa968611",
   "metadata": {},
   "source": [
    "## Create reverse complement of old sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5ff61ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"raw/Megan.nexus\"\n",
    "output_file = \"raw/Megan_revcomp.nex\"\n",
    "revcomp_records = []\n",
    "\n",
    "for seq_record in SeqIO.parse(input_file, \"nexus\"):\n",
    "    reverse_complement_seq = seq_record.seq.reverse_complement()\n",
    "    reverse_complement_record = SeqRecord(\n",
    "        reverse_complement_seq,\n",
    "        id=seq_record.id,\n",
    "        description=seq_record.description + \" reverse_complemented\",\n",
    "    )\n",
    "    reverse_complement_record.annotations[\"molecule_type\"] = \"DNA\"\n",
    "    revcomp_records.append(reverse_complement_record)\n",
    "\n",
    "# Write all at once\n",
    "with open(output_file, \"w\") as output_handle:\n",
    "    SeqIO.write(revcomp_records, output_handle, \"nexus\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09d118a5",
   "metadata": {},
   "source": [
    "## Join all sequences together and trim to overlapping region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "399727ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "316"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List your input NEXUS files\n",
    "input_files = [\"raw/Sardinia.nexus\", \"raw/Megan_revcomp.nex\", \"raw/Rasit_revcomp.nex\"]\n",
    "# input_files = [\"raw/Rasit.nex\", \"raw/Sardinia.nexus\"]\n",
    "\n",
    "# Output FASTA file\n",
    "output_file = \"processed/combined_CO1.fasta\"\n",
    "\n",
    "# Collect all sequences\n",
    "all_records = []\n",
    "for f in input_files:\n",
    "    for record in SeqIO.parse(f, \"nexus\"):\n",
    "        all_records.append(record)\n",
    "\n",
    "# Write all sequences (unaligned) to one FASTA\n",
    "SeqIO.write(all_records, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2b18c2",
   "metadata": {},
   "source": [
    "sites 106-588 common between new and old sequences  \n",
    "save aligned file using mega"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99e4363d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "316"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## attempting to trim out excess\n",
    "input_file = \"processed/combined_CO1_aligned.fas\"\n",
    "output_file = \"processed/combined_CO1_trimmed.fasta\"\n",
    "\n",
    "start = 106  # 1-based inclusive\n",
    "end = 588    # 1-based inclusive\n",
    "\n",
    "trimmed_records = []\n",
    "\n",
    "for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "    trimmed_seq = record.seq[start-1:end]\n",
    "    record.seq = trimmed_seq\n",
    "    trimmed_records.append(record)\n",
    "    \n",
    "SeqIO.write(trimmed_records, output_file, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "123737a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO 11/10:\n",
    "'''\n",
    "extract FASTA file header names for raw / combined \n",
    "    - remove Mattamuskeet duplicates, check that sequences are identical\n",
    "    - check for any other duplicates in combined file\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cddf876",
   "metadata": {},
   "source": [
    "### Extract FASTA headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "144fea71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_file = [\"raw/Rasit_revcomp.nex\"]\n",
    "output_file = \"raw/Rasit_revcomp.fasta\"\n",
    "rasit_records = []\n",
    "for f in input_file:\n",
    "    for record in SeqIO.parse(f, \"nexus\"):\n",
    "        rasit_records.append(record)\n",
    "\n",
    "# Write all sequences (unaligned) to one FASTA\n",
    "SeqIO.write(rasit_records, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abf86b10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MtD2',\n",
       " 'Tk_16',\n",
       " 'Tx_43',\n",
       " 'Tk_20',\n",
       " 'Tx_54',\n",
       " 'MtC7f',\n",
       " 'Tk_3',\n",
       " 'Tk_8',\n",
       " 'Tx_26',\n",
       " 'Tx_45',\n",
       " 'Bft_1',\n",
       " 'Bft_2',\n",
       " 'Bft_21',\n",
       " 'Bft_29',\n",
       " 'Bft_3',\n",
       " 'Bft_34',\n",
       " 'Bft_5',\n",
       " 'Bft_8',\n",
       " 'Ga_102',\n",
       " 'Ga_103',\n",
       " 'Ga_104',\n",
       " 'Ga_106',\n",
       " 'Ga_107',\n",
       " 'Ga_111',\n",
       " 'Ga_112',\n",
       " 'Ga_114',\n",
       " 'Ga_115',\n",
       " 'Ga_116',\n",
       " 'Ga_117',\n",
       " 'Ga_118',\n",
       " 'Ga_119',\n",
       " 'Ga_120',\n",
       " 'Ga_121',\n",
       " 'Ga_122',\n",
       " 'Ga_123',\n",
       " 'Ga_124',\n",
       " 'Ga_125',\n",
       " 'Ga_126',\n",
       " 'Ga_127',\n",
       " 'Ga_130',\n",
       " 'Ga_131',\n",
       " 'Ga_132',\n",
       " 'Ga_133',\n",
       " 'Ga_134',\n",
       " 'Ga_135',\n",
       " 'Ga_136',\n",
       " 'Ga_137',\n",
       " 'Ga_139',\n",
       " 'Ga_140',\n",
       " 'MtC4f',\n",
       " 'Tk_1',\n",
       " 'Tk_18',\n",
       " 'Tx_24',\n",
       " 'Tx_25',\n",
       " 'Tx_36',\n",
       " 'Tx_42',\n",
       " 'Tx_46',\n",
       " 'Tx_57',\n",
       " 'Tk_7',\n",
       " 'Tx_33',\n",
       " 'Bft_6',\n",
       " 'MtC3',\n",
       " 'Tk_12',\n",
       " 'Tk_17',\n",
       " 'Tk_9',\n",
       " 'Tx_44',\n",
       " 'Tx_48',\n",
       " 'Tx_51',\n",
       " 'Tx_53',\n",
       " 'Bft_4',\n",
       " 'Bft_13',\n",
       " 'Bft_28',\n",
       " 'Bft_7',\n",
       " 'Bft_25',\n",
       " 'Bft_27',\n",
       " 'Bft_30',\n",
       " 'Ch_11',\n",
       " 'Ch_12',\n",
       " 'Ch_14',\n",
       " 'Ch_16',\n",
       " 'Ch_17',\n",
       " 'Ch_18',\n",
       " 'Ch_19',\n",
       " 'Ch_2',\n",
       " 'Ch_20',\n",
       " 'Ch_21',\n",
       " 'Ch_23',\n",
       " 'Ch_25',\n",
       " 'Ch_26',\n",
       " 'Ch_27',\n",
       " 'Ch_28',\n",
       " 'Ch_3',\n",
       " 'Ch_30',\n",
       " 'Ch_4',\n",
       " 'Ch_6',\n",
       " 'Ch_7',\n",
       " 'Ch_9',\n",
       " 'MtC1',\n",
       " 'MtC4',\n",
       " 'MtC5f',\n",
       " 'MtC6',\n",
       " 'MtC7',\n",
       " 'MtC9f',\n",
       " 'MtD1f',\n",
       " 'Tk_14',\n",
       " 'Tx_27',\n",
       " 'Tx_28',\n",
       " 'Tx_29',\n",
       " 'Tx_31',\n",
       " 'Tx_34',\n",
       " 'Tx_38',\n",
       " 'Tx_55',\n",
       " 'Bft_12',\n",
       " 'Bft_26',\n",
       " 'Bft_33',\n",
       " 'Ch_10',\n",
       " 'Tk_26',\n",
       " 'Tk_28',\n",
       " 'Tk_29',\n",
       " 'Tk_31',\n",
       " 'Tk_32',\n",
       " 'Tk_33',\n",
       " 'Tk_36',\n",
       " 'Tk_38',\n",
       " 'Tk_44',\n",
       " 'Tk_45',\n",
       " 'Tx_30',\n",
       " 'Bft_10',\n",
       " 'Bft_14',\n",
       " 'Bft_15',\n",
       " 'Bft_20',\n",
       " 'Bft_9',\n",
       " 'MtA1',\n",
       " 'MtA3',\n",
       " 'MtA4',\n",
       " 'MtA5',\n",
       " 'MtA7',\n",
       " 'MtA9',\n",
       " 'MtB1',\n",
       " 'MtB3',\n",
       " 'MtB4',\n",
       " 'MtB7',\n",
       " 'MtC5',\n",
       " 'MtC6f',\n",
       " 'MtC8f',\n",
       " 'Tk_35',\n",
       " 'Tk_46',\n",
       " 'Tk_47',\n",
       " 'Tx_39',\n",
       " 'Tx_47',\n",
       " 'Bft_11',\n",
       " 'MtB5',\n",
       " 'MtB9',\n",
       " 'Tk_27',\n",
       " 'Tx_40',\n",
       " 'Tk_40',\n",
       " 'Tk_19',\n",
       " 'Tk_2',\n",
       " 'Tx_56',\n",
       " 'Tk_15',\n",
       " 'Tk_34',\n",
       " 'Bft_17']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_fasta_headers(fasta_file_path):\n",
    "    headers = []\n",
    "    with open(fasta_file_path, \"r\") as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            headers.append(record.description)\n",
    "    return headers\n",
    "\n",
    "rasit = \"raw/Rasit_revcomp.fasta\" \n",
    "megan = \"raw/Megan_revcomp.fasta\"\n",
    "rasit_headers = get_fasta_headers(rasit)\n",
    "megan_headers = get_fasta_headers(megan)\n",
    "\n",
    "rasit_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5ed0f91b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Headers starting with 'Mt' unique to 'all': set()\n",
      "Headers starting with 'Mt' unique to 'megan': set()\n",
      "Shared 'Mt' headers: {'MtA1', 'MtC1', 'MtB3', 'MtC4', 'MtC5f', 'MtB4', 'MtC7', 'MtA3', 'MtA4', 'MtA5', 'MtC9f', 'MtB9', 'MtC8f', 'MtC6', 'MtB1', 'MtA9', 'MtB7', 'MtC4f', 'MtC7f', 'MtD1f', 'MtC5', 'MtB5', 'MtC6f', 'MtA7', 'MtC3', 'MtD2'}\n"
     ]
    }
   ],
   "source": [
    "def get_mt_headers(fasta_file_path, clean_underscores=False):\n",
    "    \"\"\"Return only headers starting with 'Mt'.\"\"\"\n",
    "    headers = [h for h in get_fasta_headers(fasta_file_path) if h.startswith(\"Mt\")]\n",
    "    if clean_underscores:\n",
    "        headers = [h.replace(\"_\", \"\") for h in headers]\n",
    "    return headers\n",
    "\n",
    "rasit_mt_headers = set(get_mt_headers(rasit))\n",
    "megan_mt_headers = set(get_mt_headers(megan, clean_underscores=True))\n",
    "\n",
    "unique_to_rasit = rasit_mt_headers - megan_mt_headers\n",
    "unique_to_megan = megan_mt_headers - rasit_mt_headers\n",
    "shared = rasit_mt_headers & megan_mt_headers\n",
    "\n",
    "print(\"Headers starting with 'Mt' unique to 'all':\", unique_to_rasit)\n",
    "print(\"Headers starting with 'Mt' unique to 'megan':\", unique_to_megan)\n",
    "print(\"Shared 'Mt' headers:\", shared)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16c86d86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Headers starting with 'Bft' unique to 'all': set()\n",
      "Headers starting with 'Bft' unique to 'megan': {'Bft101', 'Bft121'}\n",
      "Shared 'Bft' headers: {'Bft15', 'Bft10', 'Bft34', 'Bft17', 'Bft28', 'Bft4', 'Bft33', 'Bft30', 'Bft1', 'Bft5', 'Bft21', 'Bft11', 'Bft26', 'Bft12', 'Bft7', 'Bft3', 'Bft25', 'Bft8', 'Bft2', 'Bft6', 'Bft13', 'Bft27', 'Bft29', 'Bft9', 'Bft14', 'Bft20'}\n"
     ]
    }
   ],
   "source": [
    "def get_bft_headers(fasta_file_path, clean_underscores=True):\n",
    "    \"\"\"Return only headers starting with 'Mt'.\"\"\"\n",
    "    headers = [h for h in get_fasta_headers(fasta_file_path) if h.startswith(\"Bft\")]\n",
    "    if clean_underscores:\n",
    "        headers = [h.replace(\"_\", \"\") for h in headers]\n",
    "    return headers\n",
    "\n",
    "rasit = \"raw/Rasit_revcomp.fasta\" \n",
    "megan = \"raw/Megan_revcomp.fasta\"\n",
    "\n",
    "rasit_bft_headers = set(get_bft_headers(rasit))\n",
    "megan_bft_headers = set(get_bft_headers(megan))\n",
    "\n",
    "unique_to_rasit = rasit_bft_headers - megan_bft_headers\n",
    "unique_to_megan = megan_bft_headers - rasit_bft_headers\n",
    "shared = rasit_bft_headers & megan_bft_headers\n",
    "\n",
    "print(\"Headers starting with 'Bft' unique to 'rasit':\", unique_to_rasit)\n",
    "print(\"Headers starting with 'Bft' unique to 'megan':\", unique_to_megan)\n",
    "print(\"Shared 'Bft' headers:\", shared)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39627a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO\n",
    "# remake combined nexus with correct names, no 'copy'\n",
    "# convert names into trait matrix for haplotype analysis\n",
    "# make haplotype map TCS network, send to Rasit/Dan\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a3054c",
   "metadata": {},
   "source": [
    "## Write nexus from trimmed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff295dc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"processed/combined_CO1_trimmed.fasta\"\n",
    "output_file = \"processed/combined_CO1_trimmed.nex\"\n",
    "\n",
    "records = []\n",
    "for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "    record.annotations[\"molecule_type\"] = \"DNA\"\n",
    "    records.append(record)\n",
    "\n",
    "count = SeqIO.write(records, output_file, \"nexus\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbb019be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
