{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "49ea04a1",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad05077",
   "metadata": {},
   "source": [
    "note: sequence alignment and conversion of FASTA into NEXUS was done in MEGA11 software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1683296",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/henrysun_1/Desktop/Duke/PhD 2025-2026/callinectes_co1\n"
     ]
    }
   ],
   "source": [
    "import Bio \n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import re\n",
    "\n",
    "print(os.getcwd())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa968611",
   "metadata": {},
   "source": [
    "## Create reverse complement of old sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5ff61ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"raw/Megan.nexus\"\n",
    "output_file = \"raw/Megan_revcomp.nex\"\n",
    "revcomp_records = []\n",
    "\n",
    "for seq_record in SeqIO.parse(input_file, \"nexus\"):\n",
    "    reverse_complement_seq = seq_record.seq.reverse_complement()\n",
    "    reverse_complement_record = SeqRecord(\n",
    "        reverse_complement_seq,\n",
    "        id=seq_record.id,\n",
    "        description=seq_record.description + \" reverse_complemented\",\n",
    "    )\n",
    "    reverse_complement_record.annotations[\"molecule_type\"] = \"DNA\"\n",
    "    revcomp_records.append(reverse_complement_record)\n",
    "\n",
    "# Write all at once\n",
    "with open(output_file, \"w\") as output_handle:\n",
    "    SeqIO.write(revcomp_records, output_handle, \"nexus\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09d118a5",
   "metadata": {},
   "source": [
    "## Join all sequences together and trim to overlapping region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "399727ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "316"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List your input NEXUS files\n",
    "input_files = [\"raw/Sardinia.nexus\", \"raw/Megan_revcomp.nex\", \"raw/Rasit_revcomp.nex\"]\n",
    "# input_files = [\"raw/Rasit.nex\", \"raw/Sardinia.nexus\"]\n",
    "\n",
    "# Output FASTA file\n",
    "output_file = \"processed/combined_CO1.fasta\"\n",
    "\n",
    "# Collect all sequences\n",
    "all_records = []\n",
    "for f in input_files:\n",
    "    for record in SeqIO.parse(f, \"nexus\"):\n",
    "        all_records.append(record)\n",
    "\n",
    "# Write all sequences (unaligned) to one FASTA\n",
    "SeqIO.write(all_records, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd2b18c2",
   "metadata": {},
   "source": [
    "sites 106-588 common between new and old sequences  \n",
    "save aligned file using mega"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "99e4363d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "264"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## attempting to trim out excess\n",
    "input_file = \"processed/combined_unique_aligned.fas\"\n",
    "output_file = \"processed/combined_unique_trimmed.fasta\"\n",
    "\n",
    "start = 106  # 1-based inclusive\n",
    "end = 588    # 1-based inclusive\n",
    "\n",
    "trimmed_records = []\n",
    "\n",
    "for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "    trimmed_seq = record.seq[start-1:end]\n",
    "    record.seq = trimmed_seq\n",
    "    trimmed_records.append(record)\n",
    "    \n",
    "SeqIO.write(trimmed_records, output_file, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "123737a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO 11/10:\n",
    "'''\n",
    "extract FASTA file header names for raw / combined \n",
    "    - remove Mattamuskeet duplicates, check that sequences are identical\n",
    "    - check for any other duplicates in combined file\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cddf876",
   "metadata": {},
   "source": [
    "### Aside: Extract FASTA headers and remake FASTA with unique sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "144fea71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_file = [\"raw/Rasit_revcomp.nex\"]\n",
    "output_file = \"raw/Rasit_revcomp.fasta\"\n",
    "rasit_records = []\n",
    "for f in input_file:\n",
    "    for record in SeqIO.parse(f, \"nexus\"):\n",
    "        rasit_records.append(record)\n",
    "\n",
    "# Write all sequences (unaligned) to one FASTA\n",
    "SeqIO.write(rasit_records, output_file, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abf86b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_headers(fasta_file_path):\n",
    "    headers = []\n",
    "    with open(fasta_file_path, \"r\") as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            headers.append(record.description)\n",
    "    return headers\n",
    "\n",
    "rasit = \"raw/Rasit_revcomp.fasta\" \n",
    "megan = \"raw/Megan_revcomp.fasta\"\n",
    "rasit_headers = get_fasta_headers(rasit)\n",
    "megan_headers = get_fasta_headers(megan)\n",
    "\n",
    "rasit_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5ed0f91b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Headers starting with 'Mt' unique to 'all': set()\n",
      "Headers starting with 'Mt' unique to 'megan': set()\n",
      "Shared 'Mt' headers: {'MtA1', 'MtC1', 'MtB3', 'MtC4', 'MtC5f', 'MtB4', 'MtC7', 'MtA3', 'MtA4', 'MtA5', 'MtC9f', 'MtB9', 'MtC8f', 'MtC6', 'MtB1', 'MtA9', 'MtB7', 'MtC4f', 'MtC7f', 'MtD1f', 'MtC5', 'MtB5', 'MtC6f', 'MtA7', 'MtC3', 'MtD2'}\n"
     ]
    }
   ],
   "source": [
    "def get_mt_headers(fasta_file_path, clean_underscores=False):\n",
    "    \"\"\"Return only headers starting with 'Mt'\"\"\"\n",
    "    headers = [h for h in get_fasta_headers(fasta_file_path) if h.startswith(\"Mt\")]\n",
    "    if clean_underscores:\n",
    "        headers = [h.replace(\"_\", \"\") for h in headers]\n",
    "    return headers\n",
    "\n",
    "rasit_mt_headers = set(get_mt_headers(rasit))\n",
    "megan_mt_headers = set(get_mt_headers(megan, clean_underscores=True))\n",
    "\n",
    "mt_unique_to_rasit = rasit_mt_headers - megan_mt_headers\n",
    "mt_unique_to_megan = megan_mt_headers - rasit_mt_headers\n",
    "shared_mt = rasit_mt_headers & megan_mt_headers\n",
    "\n",
    "print(\"Headers starting with 'Mt' unique to 'all':\", mt_unique_to_rasit)\n",
    "print(\"Headers starting with 'Mt' unique to 'megan':\", mt_unique_to_megan)\n",
    "print(\"Shared 'Mt' headers:\", shared_mt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "16c86d86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Headers starting with 'Bft' unique to 'rasit': set()\n",
      "Headers starting with 'Bft' unique to 'megan': {'Bft101', 'Bft121'}\n",
      "Shared 'Bft' headers: {'Bft15', 'Bft10', 'Bft34', 'Bft17', 'Bft28', 'Bft4', 'Bft33', 'Bft30', 'Bft1', 'Bft5', 'Bft21', 'Bft11', 'Bft26', 'Bft12', 'Bft7', 'Bft3', 'Bft25', 'Bft8', 'Bft2', 'Bft6', 'Bft13', 'Bft27', 'Bft29', 'Bft9', 'Bft14', 'Bft20'}\n"
     ]
    }
   ],
   "source": [
    "def get_bft_headers(fasta_file_path, clean_underscores=True):\n",
    "    \"\"\"Return only headers starting with 'Mt'.\"\"\"\n",
    "    headers = [h for h in get_fasta_headers(fasta_file_path) if h.startswith(\"Bft\")]\n",
    "    if clean_underscores:\n",
    "        headers = [h.replace(\"_\", \"\") for h in headers]\n",
    "    return headers\n",
    "\n",
    "rasit = \"raw/Rasit_revcomp.fasta\" \n",
    "megan = \"raw/Megan_revcomp.fasta\"\n",
    "\n",
    "rasit_bft_headers = set(get_bft_headers(rasit))\n",
    "megan_bft_headers = set(get_bft_headers(megan))\n",
    "\n",
    "bft_unique_to_rasit = rasit_bft_headers - megan_bft_headers\n",
    "bft_unique_to_megan = megan_bft_headers - rasit_bft_headers\n",
    "shared_bft = rasit_bft_headers & megan_bft_headers\n",
    "\n",
    "print(\"Headers starting with 'Bft' unique to 'rasit':\", bft_unique_to_rasit)\n",
    "print(\"Headers starting with 'Bft' unique to 'megan':\", bft_unique_to_megan)\n",
    "print(\"Shared 'Bft' headers:\", shared_bft)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a8a4ec68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original Rasit sequences: 162\n",
      "Filtered Rasit sequences: 110 (removed 52)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "264"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rasit_file = \"raw/Rasit_revcomp.fasta\"\n",
    "megan_file = \"raw/Megan_revcomp.fasta\"\n",
    "sardinia_file = \"raw/Sardinia.nexus\"\n",
    "rasit_unique_file = \"raw/Rasit_unique.fasta\"\n",
    "combined_unique_file = \"processed/combined_unique.fasta\"\n",
    "\n",
    "shared_headers = shared_mt | shared_bft\n",
    "rasit_records = list(SeqIO.parse(rasit_file, \"fasta\"))\n",
    "rasit_filtered = []\n",
    "\n",
    "for rec in rasit_records:\n",
    "    clean_header = rec.description.replace(\"_\", \"\")\n",
    "    if clean_header not in shared_headers:\n",
    "        rasit_filtered.append(rec)\n",
    "\n",
    "print(f\"Original Rasit sequences: {len(rasit_records)}\")\n",
    "print(f\"Filtered Rasit sequences: {len(rasit_filtered)} (removed {len(rasit_records) - len(rasit_filtered)})\")\n",
    "\n",
    "combined_records = []\n",
    "combined_records.extend(list(SeqIO.parse(rasit_unique_file, \"fasta\")))\n",
    "combined_records.extend(list(SeqIO.parse(megan_file, \"fasta\")))\n",
    "combined_records.extend(list(SeqIO.parse(sardinia_file, \"nexus\")))\n",
    "\n",
    "SeqIO.write(combined_records, combined_unique_file, \"fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a92c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file = \"processed/combined_unique.fasta\"\n",
    "\n",
    "pattern = re.compile(r'^(>C)(\\d+)')\n",
    "# also manually cf replace Bft, Sardinia with _\n",
    "\n",
    "with open(combined_unique_file) as infile, open(output_file, \"w\") as outfile:\n",
    "    for line in infile:\n",
    "        outfile.write(pattern.sub(r'\\1_\\2', line))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "7c529307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tk_16</th>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tx_43</th>\n",
       "      <td>Tx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tk_20</th>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tx_54</th>\n",
       "      <td>Tx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tk_3</th>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_7</th>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_6</th>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_4</th>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_2</th>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_1</th>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Population\n",
       "Taxon                \n",
       "Tk_16              Tk\n",
       "Tx_43              Tx\n",
       "Tk_20              Tk\n",
       "Tx_54              Tx\n",
       "Tk_3               Tk\n",
       "...               ...\n",
       "Sardinia_7   Sardinia\n",
       "Sardinia_6   Sardinia\n",
       "Sardinia_4   Sardinia\n",
       "Sardinia_2   Sardinia\n",
       "Sardinia_1   Sardinia\n",
       "\n",
       "[264 rows x 1 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## trait matrix to csv\n",
    "df = pd.read_csv(\"processed/population_traits.csv\", index_col=0)\n",
    "df\n",
    "# df.to_csv(\"processed/population_traits.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39627a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO\n",
    "# convert names into trait matrix for haplotype analysis\n",
    "# make haplotype map TCS network, send to Rasit/Dan\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4adeefe",
   "metadata": {},
   "source": [
    "for Sardinian crabs - minor note: \n",
    "four cases of - (gap in sequence)\n",
    "try filling gaps with canonical sequence from other crabs to see impact on network. could all have 1 haplotype?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dea1eb92",
   "metadata": {},
   "source": [
    "## Formatting trait matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "053165e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Texas</th>\n",
       "      <th>Georgia</th>\n",
       "      <th>Chesapeake</th>\n",
       "      <th>Carrot Island</th>\n",
       "      <th>Mattamuskeet</th>\n",
       "      <th>Sardinia</th>\n",
       "      <th>Turkey</th>\n",
       "      <th>Beaufort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Tk_16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Tx_43</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Tk_20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Tx_54</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Tk_3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>259</td>\n",
       "      <td>Sardinia_7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>260</td>\n",
       "      <td>Sardinia_6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>261</td>\n",
       "      <td>Sardinia_4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>262</td>\n",
       "      <td>Sardinia_2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>263</td>\n",
       "      <td>Sardinia_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0       Taxon  Texas  Georgia  Chesapeake  Carrot Island  \\\n",
       "0             0       Tk_16      0        0           0              0   \n",
       "1             1       Tx_43      1        0           0              0   \n",
       "2             2       Tk_20      0        0           0              0   \n",
       "3             3       Tx_54      1        0           0              0   \n",
       "4             4        Tk_3      0        0           0              0   \n",
       "..          ...         ...    ...      ...         ...            ...   \n",
       "259         259  Sardinia_7      0        0           0              0   \n",
       "260         260  Sardinia_6      0        0           0              0   \n",
       "261         261  Sardinia_4      0        0           0              0   \n",
       "262         262  Sardinia_2      0        0           0              0   \n",
       "263         263  Sardinia_1      0        0           0              0   \n",
       "\n",
       "     Mattamuskeet  Sardinia  Turkey  Beaufort  \n",
       "0               0         0       1         0  \n",
       "1               0         0       0         0  \n",
       "2               0         0       1         0  \n",
       "3               0         0       0         0  \n",
       "4               0         0       1         0  \n",
       "..            ...       ...     ...       ...  \n",
       "259             0         1       0         0  \n",
       "260             0         1       0         0  \n",
       "261             0         1       0         0  \n",
       "262             0         1       0         0  \n",
       "263             0         1       0         0  \n",
       "\n",
       "[264 rows x 10 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"processed/reformatted_traits.csv\")\n",
    "# # df2 = df2.drop(index=df2.index[0], inplace=True)\n",
    "# df2 = df.drop(columns=['Taxon'], axis=1)\n",
    "# df2['row_sum'] = df2.sum(axis=1)\n",
    "# print(df2['row_sum'])\n",
    "# print((df2['row_sum'] == 1).all())\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5a6b8ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Texas</th>\n",
       "      <th>Georgia</th>\n",
       "      <th>Chesapeake</th>\n",
       "      <th>Carrot Island</th>\n",
       "      <th>Mattamuskeet</th>\n",
       "      <th>Sardinia</th>\n",
       "      <th>Turkey</th>\n",
       "      <th>Beaufort</th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Tk_16</td>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Tx_43</td>\n",
       "      <td>Tx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Tk_20</td>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Tx_54</td>\n",
       "      <td>Tx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Tk_3</td>\n",
       "      <td>Tk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sardinia_7</td>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sardinia_6</td>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sardinia_4</td>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sardinia_2</td>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sardinia_1</td>\n",
       "      <td>Sardinia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Texas  Georgia  Chesapeake  Carrot Island  Mattamuskeet  Sardinia  \\\n",
       "0        0        0           0              0             0         0   \n",
       "1        1        0           0              0             0         0   \n",
       "2        0        0           0              0             0         0   \n",
       "3        1        0           0              0             0         0   \n",
       "4        0        0           0              0             0         0   \n",
       "..     ...      ...         ...            ...           ...       ...   \n",
       "259      0        0           0              0             0         1   \n",
       "260      0        0           0              0             0         1   \n",
       "261      0        0           0              0             0         1   \n",
       "262      0        0           0              0             0         1   \n",
       "263      0        0           0              0             0         1   \n",
       "\n",
       "     Turkey  Beaufort       Taxon Population  \n",
       "0         1         0       Tk_16         Tk  \n",
       "1         0         0       Tx_43         Tx  \n",
       "2         1         0       Tk_20         Tk  \n",
       "3         0         0       Tx_54         Tx  \n",
       "4         1         0        Tk_3         Tk  \n",
       "..      ...       ...         ...        ...  \n",
       "259       0         0  Sardinia_7   Sardinia  \n",
       "260       0         0  Sardinia_6   Sardinia  \n",
       "261       0         0  Sardinia_4   Sardinia  \n",
       "262       0         0  Sardinia_2   Sardinia  \n",
       "263       0         0  Sardinia_1   Sardinia  \n",
       "\n",
       "[264 rows x 10 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"processed/reformatted_traits.csv\",index_col=0)\n",
    "df2 = pd.read_csv(\"processed/population_traits.csv\")\n",
    "df = pd.concat([df, df2], axis=1)\n",
    "\n",
    "# df = df.rename(columns={'Tx': 'Texas', 'Ga': 'Georgia','Ch': 'Chesapeake','C': 'Carrot Island','Mt': 'Mattamuskeet','Tk': 'Turkey', 'Bft': 'Beaufort'})\n",
    "# df = df.drop(columns=['Unnamed: 0'], axis=1)\n",
    "# df.to_csv(\"processed/reformatted_traits.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48918b95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Texas</th>\n",
       "      <th>Georgia</th>\n",
       "      <th>Chesapeake</th>\n",
       "      <th>Carrot Island</th>\n",
       "      <th>Mattamuskeet</th>\n",
       "      <th>Sardinia</th>\n",
       "      <th>Turkey</th>\n",
       "      <th>Beaufort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tk_16</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tx_43</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tk_20</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tx_54</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tk_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sardinia_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Texas  Georgia  Chesapeake  Carrot Island  Mattamuskeet  Sardinia  \\\n",
       "Tk_16           0        0           0              0             0         0   \n",
       "Tx_43           1        0           0              0             0         0   \n",
       "Tk_20           0        0           0              0             0         0   \n",
       "Tx_54           1        0           0              0             0         0   \n",
       "Tk_3            0        0           0              0             0         0   \n",
       "...           ...      ...         ...            ...           ...       ...   \n",
       "Sardinia_7      0        0           0              0             0         1   \n",
       "Sardinia_6      0        0           0              0             0         1   \n",
       "Sardinia_4      0        0           0              0             0         1   \n",
       "Sardinia_2      0        0           0              0             0         1   \n",
       "Sardinia_1      0        0           0              0             0         1   \n",
       "\n",
       "            Turkey  Beaufort  \n",
       "Tk_16            1         0  \n",
       "Tx_43            0         0  \n",
       "Tk_20            1         0  \n",
       "Tx_54            0         0  \n",
       "Tk_3             1         0  \n",
       "...            ...       ...  \n",
       "Sardinia_7       0         0  \n",
       "Sardinia_6       0         0  \n",
       "Sardinia_4       0         0  \n",
       "Sardinia_2       0         0  \n",
       "Sardinia_1       0         0  \n",
       "\n",
       "[264 rows x 8 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # df = df.set_index('Taxon')\n",
    "# df = df.drop(columns=['Population'], axis=1)\n",
    "df.index.name = None\n",
    "df.to_csv(\"processed/reformatted_traits.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "717610eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
